src/Raven.Client/Documents/DocumentStore.cs

```diff
@@ -344,24 +345,74 @@ public override IDisposable AggressivelyCacheFor(TimeSpan cacheDuration, string
         /// without touching the server.
         /// </remarks>
         public override IDisposable AggressivelyCacheFor(TimeSpan cacheDuration, AggressiveCacheMode mode, string database = null)
+        {
+            var release = SetAggressiveCache(cacheDuration, mode, ref database);
+
+            return AsyncHelpers.RunSync(() => FinalizeAggressiveCacheAsync(release, mode, database));
+        }
+
+        /// <summary>
+        /// Setup the context for aggressive caching.
+        /// </summary>
+        /// <remarks>
+        /// Aggressive caching means that we will not check the server to see whether the response
+        /// we provide is current or not, but will serve the information directly from the local cache
+        /// without touching the server.
+        /// </remarks>
+        public override Task<IDisposable> AggressivelyCacheForAsync(TimeSpan cacheDuration, string database = null)
+        {
+            return AggressivelyCacheForAsync(cacheDuration, Conventions.AggressiveCache.Mode, database);
+        }
+
+        /// <summary>
+        /// Setup the context for aggressive caching.
+        /// </summary>
+        /// <remarks>
+        /// Aggressive caching means that we will not check the server to see whether the response
+        /// we provide is current or not, but will serve the information directly from the local cache
+        /// without touching the server.
+        /// </remarks>
+        public override Task<IDisposable> AggressivelyCacheForAsync(TimeSpan cacheDuration, AggressiveCacheMode mode, string database = null)
+        {
+            var release = SetAggressiveCache(cacheDuration, mode, ref database);
+
+            return FinalizeAggressiveCacheAsync(release, mode, database);
+        }
+
+        private IDisposable SetAggressiveCache(TimeSpan cacheDuration, AggressiveCacheMode mode, ref string database)
         {
             AssertInitialized();
 
             database = this.GetDatabase(database);
 
-            if (mode != AggressiveCacheMode.DoNotTrackChanges)
-                ListenToChangesAndUpdateTheCache(database);
+            var requestExecutor = GetRequestExecutor(database);
+            var oldValue = requestExecutor.AggressiveCaching.Value;
+            var newValue = new AggressiveCacheOptions(cacheDuration, mode);
 
-            var re = GetRequestExecutor(database);
-            var old = re.AggressiveCaching.Value;
-            var @new = new AggressiveCacheOptions(cacheDuration, mode);
+            requestExecutor.AggressiveCaching.Value = newValue;
 
-            re.AggressiveCaching.Value = @new;
+            return new DisposableAction(() => requestExecutor.AggressiveCaching.Value = oldValue);
```

Comment:
Is there a reason why the action doesn't provide a state-based overload that would avoid the closure?